src/main/java/com/puppycrawl/tools/checkstyle/checks/design/FinalClassCheck.java

```diff
@@ -168,6 +173,19 @@ public void visitToken(DetailAST ast) {
                 }
                 break;
 
+            case TokenTypes.LITERAL_NEW:
+                if (ast.getFirstChild() != null) {
+                    final ClassDesc currClass = classes.peek();
+                    final String qualifiedAnonInnerClassName =
+                            packageName + PACKAGE_SEPARATOR + ast.getFirstChild().getText();
+                    if (currClass != null
+                            && qualifiedAnonInnerClassName.equals(currClass.getQualifiedName())
```

Comment:
Why is this not just `ast.getFirstChild().getText().equals(currClass.getName)` ?

I believe you can't import a 2nd class with the same name as the main class. In addition, if a anonymous class is nested inside itself, you it doesn't matter how nested it is unless it matches the current class name.

Reply:
`currClass.getQualifiedName()` is something like `com.puupycrawl.tools.ClassA` whereas `ast.getFirstChild().getText().equals(currClass.getName)` is equal to `ClassA`. I did some changes, so now it is fine imo.